src/main/java/server/ExpLogger.java

```diff
@@ -0,0 +1,93 @@
+package server;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadFactory;
+import java.sql.Timestamp;
+import static java.util.concurrent.TimeUnit.*;
+
+import config.YamlConfig;
+import tools.DatabaseConnection;
+
+public class ExpLogger {
+    private static final LinkedBlockingQueue<ExpLogRecord> expLoggerQueue = new LinkedBlockingQueue<>();
+    private static final short EXP_LOGGER_THREAD_SLEEP_DURATION_SECONDS = 60;
+    private static final short EXP_LOGGER_THREAD_SHUTDOWN_WAIT_DURATION_MINUTES = 5;
+
+    public record ExpLogRecord(int worldExpRate, int expCoupon, long gainedExp, int currentExp,Timestamp expGainTime, int charid) {}
+
+    public static void putExpLogRecord(ExpLogRecord expLogRecord) {
+        try {
+            expLoggerQueue.put(expLogRecord);
```

Comment:
Memory leak if the feature is disabled. The queue will continue to grow over time since nothing is consuming the records.

Reply:
The `putExpLogRecord` method is inside the `Character.java` -> `gainExpInternal` method:
`
if (YamlConfig.config.server.USE_EXP_GAIN_LOG && level != getMaxLevel()) {
                  ...
                    ExpLogger.putExpLogRecord(expLogRecord);
                }
`
currently shouldn't cause a memory leak, need to know if you want to resolve it in a different way

Reply:
Oh my bad, I missed it. Your solution works well, no need to change it. 
I'll check out your branch and do some testing before approval.